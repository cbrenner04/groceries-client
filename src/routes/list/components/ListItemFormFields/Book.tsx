import React, { type ChangeEventHandler } from 'react';
import { Row } from 'react-bootstrap';

import { CategoryField, TextField, NumberField, CheckboxField } from 'components/FormFields';

export interface IBookFormFieldsProps {
  author: string;
  title: string;
  purchased: boolean;
  read: boolean;
  editForm: boolean;
  numberInSeries?: number;
  category: string;
  categories: string[];
  inputChangeHandler: ChangeEventHandler;
}

const BookFormFields: React.FC<IBookFormFieldsProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <TextField
      name="author"
      label="Author"
      value={props.author}
      handleChange={props.inputChangeHandler}
      placeholder="Kurt Vonnagut"
    />
    <TextField
      name="title"
      label="Title"
      value={props.title}
      handleChange={props.inputChangeHandler}
      placeholder="Slaughterhouse-Five"
    />
    <NumberField
      name="number_in_series"
      label="Number in series"
      value={props.numberInSeries ?? 0}
      handleChange={props.inputChangeHandler}
    />
    <CategoryField category={props.category} categories={props.categories} handleInput={props.inputChangeHandler} />
    {props.editForm && (
      <Row className="mb-3">
        <CheckboxField
          name="purchased"
          label="Purchased"
          value={props.purchased}
          handleChange={props.inputChangeHandler}
          classes="form-check-inline ms-1"
        />
        <CheckboxField
          name="read"
          label="Read"
          value={props.read}
          handleChange={props.inputChangeHandler}
          classes="form-check-inline"
        />
      </Row>
    )}
  </React.Fragment>
);

export default BookFormFields;
