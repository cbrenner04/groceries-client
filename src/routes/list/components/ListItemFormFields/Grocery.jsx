import React from 'react';
import PropTypes from 'prop-types';

import { CategoryField, TextField, CheckboxField } from '../../../../components/FormFields';

const GroceryFormFields = (props) => (
  <>
    <TextField
      name="product"
      label="Product"
      value={props.product}
      handleChange={props.inputChangeHandler}
      placeholder="apples"
    />
    <TextField
      name="quantity"
      label="Quantity"
      value={props.quantity}
      handleChange={props.inputChangeHandler}
      placeholder="3 bags"
    />
    <CategoryField
      category={props.category || ''}
      categories={props.categories || []}
      handleInput={props.inputChangeHandler}
    />
    {props.editForm && (
      <CheckboxField
        name="purchased"
        label="Purchased"
        value={props.purchased || false}
        handleChange={props.inputChangeHandler}
        classes="mb-3"
      />
    )}
  </>
);

GroceryFormFields.propTypes = {
  product: PropTypes.string.isRequired,
  quantity: PropTypes.string.isRequired,
  purchased: PropTypes.bool,
  editForm: PropTypes.bool,
  category: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
  inputChangeHandler: PropTypes.func.isRequired,
};

export default GroceryFormFields;
