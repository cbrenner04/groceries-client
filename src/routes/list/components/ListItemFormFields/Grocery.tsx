import React, { type ChangeEventHandler } from 'react';

import { CategoryField, TextField, CheckboxField } from 'components/FormFields';

export interface IGroceryFormFieldsProps {
  product: string;
  quantity: string;
  purchased: boolean;
  editForm: boolean;
  category: string;
  categories: string[];
  inputChangeHandler: ChangeEventHandler;
}
const GroceryFormFields: React.FC<IGroceryFormFieldsProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <TextField
      name="product"
      label="Product"
      value={props.product}
      handleChange={props.inputChangeHandler}
      placeholder="apples"
    />
    <TextField
      name="quantity"
      label="Quantity"
      value={props.quantity}
      handleChange={props.inputChangeHandler}
      placeholder="3 bags"
    />
    <CategoryField category={props.category} categories={props.categories} handleInput={props.inputChangeHandler} />
    {props.editForm && (
      <CheckboxField
        name="purchased"
        label="Purchased"
        value={props.purchased}
        handleChange={props.inputChangeHandler}
        classes="mb-3"
      />
    )}
  </React.Fragment>
);

export default GroceryFormFields;
