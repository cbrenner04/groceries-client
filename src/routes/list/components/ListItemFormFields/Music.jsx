import React from 'react';
import PropTypes from 'prop-types';

import { TextField, CategoryField, CheckboxField } from '../../../../components/FormFields';

const MusicFormFields = (props) => (
  <>
    <TextField
      name="title"
      label="Title"
      value={props.title}
      handleChange={props.inputChangeHandler}
      placeholder="Baby Got Back"
    />
    <TextField
      name="artist"
      label="Artist"
      value={props.artist}
      handleChange={props.inputChangeHandler}
      placeholder="Sir Mix-a-Lot"
    />
    <TextField
      name="album"
      label="Album"
      value={props.album}
      handleChange={props.inputChangeHandler}
      placeholder="Mack Daddy"
    />
    <CategoryField
      category={props.category || ''}
      categories={props.categories || []}
      handleInput={props.inputChangeHandler}
    />
    {props.editForm && (
      <CheckboxField
        name="purchased"
        label="Purchased"
        value={props.purchased || false}
        handleChange={props.inputChangeHandler}
        classes="mb-3"
      />
    )}
  </>
);

MusicFormFields.propTypes = {
  title: PropTypes.string.isRequired,
  artist: PropTypes.string.isRequired,
  album: PropTypes.string.isRequired,
  purchased: PropTypes.bool,
  editForm: PropTypes.bool,
  category: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
  inputChangeHandler: PropTypes.func.isRequired,
};

export default MusicFormFields;
