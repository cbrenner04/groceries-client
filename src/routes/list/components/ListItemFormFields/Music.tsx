import React, { ChangeEventHandler } from 'react';

import { TextField, CategoryField, CheckboxField } from '../../../../components/FormFields';

interface IMusicFormFieldsProps {
  title: string;
  artist: string;
  album: string;
  purchased?: boolean;
  editForm?: boolean;
  category?: string;
  categories?: string[];
  inputChangeHandler: ChangeEventHandler;
}

const MusicFormFields: React.FC<IMusicFormFieldsProps> = (props) => (
  <>
    <TextField
      name="title"
      label="Title"
      value={props.title}
      handleChange={props.inputChangeHandler}
      placeholder="Baby Got Back"
    />
    <TextField
      name="artist"
      label="Artist"
      value={props.artist}
      handleChange={props.inputChangeHandler}
      placeholder="Sir Mix-a-Lot"
    />
    <TextField
      name="album"
      label="Album"
      value={props.album}
      handleChange={props.inputChangeHandler}
      placeholder="Mack Daddy"
    />
    <CategoryField
      category={props.category ?? ''}
      categories={props.categories ?? []}
      handleInput={props.inputChangeHandler}
    />
    {props.editForm && (
      <CheckboxField
        name="purchased"
        label="Purchased"
        value={props.purchased ?? false}
        handleChange={props.inputChangeHandler}
        classes="mb-3"
      />
    )}
  </>
);

export default MusicFormFields;
