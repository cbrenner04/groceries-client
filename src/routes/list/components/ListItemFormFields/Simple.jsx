import React from 'react';
import PropTypes from 'prop-types';

import { TextField, CategoryField, CheckboxField } from '../../../../components/FormFields';

const SimpleFormFields = (props) => (
  <>
    <TextField
      name="content"
      label="Content"
      value={props.content}
      handleChange={props.inputChangeHandler}
      placeholder="Something cool"
    />
    <CategoryField category={props.category} categories={props.categories} handleInput={props.inputChangeHandler} />
    {props.editForm && (
      <CheckboxField
        name="completed"
        label="Completed"
        value={props.completed}
        handleChange={props.inputChangeHandler}
        classes="mb-3"
      />
    )}
  </>
);

SimpleFormFields.propTypes = {
  content: PropTypes.string.isRequired,
  completed: PropTypes.bool,
  editForm: PropTypes.bool,
  category: PropTypes.string,
  categories: PropTypes.arrayOf(PropTypes.string),
  inputChangeHandler: PropTypes.func.isRequired,
};

SimpleFormFields.defaultProps = {
  completed: false,
  editForm: false,
  category: '',
  categories: [],
};

export default SimpleFormFields;
