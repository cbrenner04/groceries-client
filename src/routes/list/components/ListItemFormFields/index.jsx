import React from 'react';
import PropTypes from 'prop-types';

import Book from './Book';
import Grocery from './Grocery';
import Music from './Music';
import Simple from './Simple';
import ToDo from './ToDo';
import { listUsers } from '../../../../types';

// TODO: reduce redundancy
const ListItemFormFields = ({ categories, listType, listUsers, formData, setFormData, editForm }) =>
  ({
    BookList: (
      <Book
        author={formData.author || ''}
        title={formData.title || ''}
        numberInSeries={formData.numberInSeries || 0}
        category={formData.category || ''}
        categories={categories || []}
        inputChangeHandler={setFormData}
        read={formData.read || false}
        purchased={formData.purchased || ''}
        editForm={editForm || false}
      />
    ),
    GroceryList: (
      <Grocery
        quantity={formData.quantity || ''}
        product={formData.product || ''}
        category={formData.category || ''}
        categories={categories || []}
        inputChangeHandler={setFormData}
        purchased={formData.purchased || ''}
        editForm={editForm || false}
      />
    ),
    MusicList: (
      <Music
        title={formData.title || ''}
        artist={formData.artist || ''}
        album={formData.album || ''}
        category={formData.category || ''}
        categories={categories || []}
        inputChangeHandler={setFormData}
        purchased={formData.purchased || ''}
        editForm={editForm || false}
      />
    ),
    SimpleList: (
      <Simple
        content={formData.content || ''}
        category={formData.category || ''}
        categories={categories || []}
        inputChangeHandler={setFormData}
        completed={formData.completed || false}
        editForm={editForm || false}
      />
    ),
    ToDoList: (
      <ToDo
        task={formData.task || ''}
        assigneeId={formData.assigneeId || ''}
        listUsers={listUsers || []}
        dueBy={formData.dueBy || ''}
        category={formData.category || ''}
        categories={categories || []}
        inputChangeHandler={setFormData}
        completed={formData.completed || false}
        editForm={editForm || false}
      />
    ),
  })[listType];

ListItemFormFields.propTypes = {
  listType: PropTypes.string.isRequired,
  listUsers: PropTypes.arrayOf(listUsers),
  categories: PropTypes.arrayOf(PropTypes.string),
  formData: PropTypes.shape({
    product: PropTypes.string,
    task: PropTypes.string,
    content: PropTypes.string,
    quantity: PropTypes.string,
    author: PropTypes.string,
    title: PropTypes.string,
    artist: PropTypes.string,
    album: PropTypes.string,
    assigneeId: PropTypes.string,
    dueBy: PropTypes.string,
    numberInSeries: PropTypes.number,
    category: PropTypes.string,
    read: PropTypes.bool,
    purchased: PropTypes.bool,
    completed: PropTypes.bool,
  }),
  setFormData: PropTypes.func.isRequired,
  editForm: PropTypes.bool,
};

export default ListItemFormFields;
