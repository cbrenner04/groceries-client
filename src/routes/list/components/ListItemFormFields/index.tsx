import React, { type ChangeEventHandler } from 'react';

import type { IListUser } from 'typings';

import Book from './Book';
import Grocery from './Grocery';
import Music from './Music';
import Simple from './Simple';
import ToDo from './ToDo';

// TODO: can this just be IListItem - I think this is confusing shit around the app
export interface IListITemsFormFieldsFormDataProps {
  id?: string;
  product?: string;
  task?: string;
  content?: string;
  quantity?: string;
  author?: string;
  title?: string;
  artist?: string;
  album?: string;
  assigneeId?: string;
  dueBy?: string;
  numberInSeries?: number;
  category?: string;
  read?: boolean;
  purchased?: boolean;
  completed?: boolean;
}

export interface IListItemFormFieldsProps {
  listType: string;
  listUsers?: IListUser[];
  categories?: string[];
  formData: IListITemsFormFieldsFormDataProps;
  setFormData: ChangeEventHandler;
  editForm?: boolean;
}

// TODO: reduce redundancy
const ListItemFormFields: React.FC<IListItemFormFieldsProps> = ({
  categories,
  listType,
  listUsers,
  formData,
  setFormData,
  editForm,
}): React.JSX.Element =>
  ({
    BookList: (
      <Book
        author={formData.author ?? ''}
        title={formData.title ?? ''}
        numberInSeries={formData.numberInSeries ?? 0}
        category={formData.category ?? ''}
        categories={categories ?? []}
        inputChangeHandler={setFormData}
        read={formData.read ?? false}
        purchased={formData.purchased ?? false}
        editForm={editForm ?? false}
      />
    ),
    GroceryList: (
      <Grocery
        quantity={formData.quantity ?? ''}
        product={formData.product ?? ''}
        category={formData.category ?? ''}
        categories={categories ?? []}
        inputChangeHandler={setFormData}
        purchased={formData.purchased ?? false}
        editForm={editForm ?? false}
      />
    ),
    MusicList: (
      <Music
        title={formData.title ?? ''}
        artist={formData.artist ?? ''}
        album={formData.album ?? ''}
        category={formData.category ?? ''}
        categories={categories ?? []}
        inputChangeHandler={setFormData}
        purchased={formData.purchased ?? false}
        editForm={editForm ?? false}
      />
    ),
    SimpleList: (
      <Simple
        content={formData.content ?? ''}
        category={formData.category ?? ''}
        categories={categories ?? []}
        inputChangeHandler={setFormData}
        completed={formData.completed ?? false}
        editForm={editForm ?? false}
      />
    ),
    ToDoList: (
      <ToDo
        task={formData.task ?? ''}
        assigneeId={formData.assigneeId ?? ''}
        listUsers={listUsers ?? []}
        dueBy={formData.dueBy ?? ''}
        category={formData.category ?? ''}
        categories={categories ?? []}
        inputChangeHandler={setFormData}
        completed={formData.completed ?? false}
        editForm={editForm ?? false}
      />
    ),
  })[listType]!;

export default ListItemFormFields;
