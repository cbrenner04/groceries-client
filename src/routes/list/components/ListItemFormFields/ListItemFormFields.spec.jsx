import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import ListItemFormFields from './index';

describe('ListItemFormFields', () => {
  const props = {
    setFormData: jest.fn(),
    formData: {
      product: '',
      task: '',
      quantity: '',
      author: '',
      title: '',
      artist: '',
      album: '',
      assigneeId: '',
      dueBy: '2020-05-21',
      numberInSeries: 0,
      category: '',
      read: false,
      purchased: false,
      completed: false,
    },
  };

  it('render Book fields when listType is BookList and calls setFormData when input is changed', () => {
    props.listType = 'BookList';
    const { container, getByLabelText } = render(<ListItemFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Author')).toBeTruthy();

    fireEvent.change(getByLabelText('Author'), { target: { value: 'a' } });

    expect(props.setFormData).toHaveBeenCalled();
  });

  it('render Grocery fields when listType is GroceryList and calls setFormData when input is changed', () => {
    props.listType = 'GroceryList';
    const { container, getByLabelText } = render(<ListItemFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Product')).toBeTruthy();

    fireEvent.change(getByLabelText('Product'), { target: { value: 'a' } });

    expect(props.setFormData).toHaveBeenCalled();
  });

  it('render Music fields when listType is MusicList and calls setFormData when input is changed', () => {
    props.listType = 'MusicList';
    const { container, getByLabelText } = render(<ListItemFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Album')).toBeTruthy();

    fireEvent.change(getByLabelText('Album'), { target: { value: 'a' } });

    expect(props.setFormData).toHaveBeenCalled();
  });

  it('render ToDo fields when listType is ToDoList and calls setFormData when input is changed', () => {
    props.listType = 'ToDoList';
    const { container, getByLabelText } = render(<ListItemFormFields {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Task')).toBeTruthy();

    fireEvent.change(getByLabelText('Task'), { target: { value: 'a' } });

    expect(props.setFormData).toHaveBeenCalled();
  });

  it('sets value for numberInSeries as a number when input', () => {
    props.listType = 'BookList';
    const { getByLabelText } = render(<ListItemFormFields {...props} />);

    fireEvent.change(getByLabelText('Number in series'), { target: { value: '2' } });

    expect(props.setFormData).toHaveBeenCalled();
  });
});
