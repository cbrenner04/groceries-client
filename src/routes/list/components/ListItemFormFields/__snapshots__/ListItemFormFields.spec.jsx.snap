// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemFormFields render Book fields when listType is BookList and calls setFormData when input is changed 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="author"
    >
      Author
    </label>
    <input
      class="form-control"
      id="author"
      name="author"
      placeholder="Kurt Vonnagut"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="title"
    >
      Title
    </label>
    <input
      class="form-control"
      id="title"
      name="title"
      placeholder="Slaughterhouse-Five"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="numberInSeries"
    >
      Number in series
    </label>
    <input
      class="form-control"
      id="numberInSeries"
      name="numberInSeries"
      type="number"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
    >
      Category
    </label>
    <input
      class="form-control"
      list="categories"
      name="category"
      type="text"
      value=""
    />
    <datalist
      id="categories"
    />
  </div>
</div>
`;

exports[`ListItemFormFields render Grocery fields when listType is GroceryList and calls setFormData when input is changed 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="product"
    >
      Product
    </label>
    <input
      class="form-control"
      id="product"
      name="product"
      placeholder="apples"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="quantity"
    >
      Quantity
    </label>
    <input
      class="form-control"
      id="quantity"
      name="quantity"
      placeholder="3 bags"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
    >
      Category
    </label>
    <input
      class="form-control"
      list="categories"
      name="category"
      type="text"
      value=""
    />
    <datalist
      id="categories"
    />
  </div>
</div>
`;

exports[`ListItemFormFields render Music fields when listType is MusicList and calls setFormData when input is changed 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="title"
    >
      Title
    </label>
    <input
      class="form-control"
      id="title"
      name="title"
      placeholder="Baby Got Back"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="artist"
    >
      Artist
    </label>
    <input
      class="form-control"
      id="artist"
      name="artist"
      placeholder="Sir Mix-a-Lot"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="album"
    >
      Album
    </label>
    <input
      class="form-control"
      id="album"
      name="album"
      placeholder="Mack Daddy"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
    >
      Category
    </label>
    <input
      class="form-control"
      list="categories"
      name="category"
      type="text"
      value=""
    />
    <datalist
      id="categories"
    />
  </div>
</div>
`;

exports[`ListItemFormFields render ToDo fields when listType is ToDoList and calls setFormData when input is changed 1`] = `
<div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="task"
    >
      Task
    </label>
    <input
      class="form-control"
      id="task"
      name="task"
      placeholder="Clean the toilets"
      type="text"
      value=""
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="assigneeId"
    >
      Assignee
    </label>
    <select
      class="form-control"
      id="assigneeId"
      name="assigneeId"
    >
      <option
        disabled=""
        value=""
      >
        Select Assignee
      </option>
    </select>
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
      for="dueBy"
    >
      Due By
    </label>
    <input
      class="form-control"
      id="dueBy"
      name="dueBy"
      placeholder="mm/dd/yyyy"
      type="date"
      value="2020-05-21"
    />
  </div>
  <div
    class="mb-3"
  >
    <label
      class="form-label"
    >
      Category
    </label>
    <input
      class="form-control"
      list="categories"
      name="category"
      type="text"
      value=""
    />
    <datalist
      id="categories"
    />
  </div>
</div>
`;
